File: Genetic_Algorithm-3/FFNNtes.py

import numpy as np
import json

with open("FFNN_terbaik.json","r") as f:
    model = json.load(f)

def decodeModelFFNN(data):
    n = 3
    input_dim = 3

    # Decode sama seperti MATLAB
    Wih = np.array(data[0 : n*input_dim]).reshape(n, input_dim)
    bih = np.array(data[n*input_dim : n*input_dim + n])
    Who = np.array(data[n*input_dim + n : n*input_dim + n + n])
    bho = data[-1]

    model = {
        "Wih": Wih.tolist(),
        "bih": bih.tolist(),
        "Who": Who.tolist(),
        "bho": bho,
        "c": 1
    }
    return model





def sigmoid(x, c=1):
    return 1 / (1 + np.exp(-c * x))

def ffnn_predict(model, x_input):
    """
    model: dict dengan kunci:
        - 'Wih' : bobot input → hidden (n x 3)
        - 'bih' : bias hidden (n)
        - 'Who' : bobot hidden → output (n)
        - 'bho' : bias output (scalar)
        - 'c'   : konstanta sigmoid
        
    x_input: array/list panjang 3
    """

    x = np.array(x_input).astype(float)  # 3 input

    Wih = np.array(model["Wih"])
    bih = np.array(model["bih"])
    Who = np.array(model["Who"])
    bho = float(model["bho"])
    c   = float(model.get("c", 1))  # default 1 kalau tidak ada

    # --- Hidden layer ---
    # SumWlb(ii) = Σ (Wih(ii,jj) * IM(jj)) + bih(ii)
    sum_hidden = Wih @ x + bih
    Xih = sigmoid(sum_hidden, c)

    # --- Output layer ---
    sum_output = np.dot(Who, Xih) + bho
    Xho = sigmoid(sum_output, c)

    return Xho

def test_interaktif(model):
    while True:
        print("\nPengujian FFNN")
        print("Masukkan tiga input biner untuk FFNN:")

        x1 = float(input("X1: "))
        x2 = float(input("X2: "))
        x3 = float(input("X3: "))

        out = ffnn_predict(model, [x1, x2, x3])
        print(f"\nOutput FFNN = {out:.6f}")

        ulang = input("Tes lagi? (y/n): ").lower()
        if ulang != 'y':
            break

if __name__ == "__main__":
    model = decodeModelFFNN(model)
    test_interaktif(model)

File: TA/FFNNtes.m
c = 1
n = 3;
n2 = n^2;

load FFNNterbaik.mat

MT = 1;
while MT == 1,
    disp('Pengujian FFNN');
    disp('');
    disp('Masukan tiga input biner unutk FFNN');
    x1 = input('X1: ');
    x2 = input('X2: ');
    x3 = input('X3: ');

    IM = [x1 x2 x3];

    NNS = FFNNterbaik;

    Wtemp = NNS(1:n2);

    for ii = 1:n,
        Wih(ii,:) = Wtemp((ii-1)*n+1:ii*n);
    end

    bih = NNS(n2+1:n2+n);
    Who = NNS(n2+n+1:n2+2*n);
    bho = NNS((n+1)^2);

    for ii=1:n,
        SumWIb = 0;
        for jj = 1:n,
            SumWIb = SumWIb + Wih(ii,jj) * IM(jj) + bih(jj);
        end
        Xih(ii) = 1/(1+exp(-c*SumWIb));
    end

    SumWXb = 0;
    for jj = 1:n,
        SumWXb  = SumWXb + Who(jj) * Xih(jj) + bho;
    end
    Xho = 1/(1+exp(-c*SumWXb));

    if Xho <= 0.02,
        HasilTes = '0';
    else 
        if Xho >= 0.98,
            HasilTes = '1';
        else 
            HasilTes = 'Unidentified';
        end
    end

    disp('');
    disp(['Output bilanagn real: ', num2str(Xho)]);
    disp(['Output FFNN adalah: ', HasilTes]);
    disp('');

    MT =input('Tekan 1 untuk tes lagi, dan tekan 0 untuk keluar:');

end